File: benchmark/http/go-fiber/internal/wrkload/service.go
package wrkload

import (
	"context"
	"fmt"
	"net/http"
	"sync"
	"time"
)

type WrkloadService interface {
	Wrk(ctx context.Context, apiData *ApiData,  connections uint, duration uint) (*WrkResult, error)
}

type wrkloadService struct {

}

func NewWrkLoadService() WrkloadService {
	return &wrkloadService{}
}

func (s *wrkloadService)Wrk(ctx context.Context, apiData *ApiData, connections uint, duration uint) (*WrkResult, error) {
	
	var wgWorker sync.WaitGroup
	var mu sync.Mutex
	chStatusCodes := make(chan int)
	successCodes := make(map[int]int)
	failureCodes := make(map[int]int)
	totalHits, failureHits := 0, 0
	start := time.Now()
	
	client := http.Client{
		Timeout: 2 * time.Second,
		Transport: &http.Transport{
			MaxIdleConns:        1000,
			MaxIdleConnsPerHost: int(connections),
			IdleConnTimeout:     90 * time.Second,
		},
	}

	ctx, cancel := context.WithTimeout(ctx, time.Duration(duration)*time.Second)
	defer cancel()

	for range connections {
		wgWorker.Go(func() {
			Worker(ctx, &client, apiData, &totalHits, &failureHits, &mu, chStatusCodes)
		})
	}

	go func() {
		wgWorker.Wait()
		close(chStatusCodes)
	}()

	for code := range chStatusCodes {
		if code <= 299 {
			successCodes[code]++
		} else {
			failureCodes[code]++
		}
	}

	return &WrkResult{
		Url: apiData.Url,
		Method: apiData.Method,
		Connections: int(connections),
		TimeTaken: fmt.Sprintf("%.3fs", time.Since(start).Seconds()),
		TotalHits: totalHits,
		SuccessHits: totalHits - failureHits,
		FailureHits: failureHits,
		SuccessMessages: successCodes,
		FailureMessages: failureCodes,
	}, nil

}

func Worker(ctx context.Context, client *http.Client, apiData *ApiData, totalHits *int, failureHits *int, mu *sync.Mutex, chStatusCodes chan<- int) {
	cntTotal, cntFailed := 0, 0
	for {
		select {
		case <-ctx.Done():
			mu.Lock()
			*totalHits += cntTotal
			*failureHits += cntFailed
			mu.Unlock()
			return
		default:
			resp, err := client.Get(apiData.Url)
			cntTotal++
			if err != nil || ( resp != nil && resp.StatusCode >= 400) {
				cntFailed++
			}
			if resp != nil {
				chStatusCodes <- resp.StatusCode
				resp.Body.Close()
			}
		}
	}
	
}

File: benchmark/http/fastapi/app/service.py
import asyncio
import time
from typing import Dict
import httpx

from app.models import ApiData, WrkResult


class WrkloadService:

    async def wrk(
        self,
        api_data: ApiData,
        connections: int,
        duration: int
    ) -> WrkResult:

        success_codes: Dict[int, int] = {}
        failure_codes: Dict[int, int] = {}

        total_hits = 0
        failure_hits = 0

        start = time.time()

        lock = asyncio.Lock()
        stop_time = start + duration

        async with httpx.AsyncClient(timeout=1.0, verify=False) as client:

            async def worker():
                nonlocal total_hits, failure_hits

                local_total = 0
                local_failed = 0

                while time.time() < stop_time:
                    try:
                        response = await client.get(api_data.url)

                        status_code = response.status_code

                        async with lock:
                            if status_code <= 299:
                                success_codes[status_code] = success_codes.get(status_code, 0) + 1
                            else:
                                failure_codes[status_code] = failure_codes.get(status_code, 0) + 1

                        local_total += 1

                    except Exception as e:
                        print(e)
                        local_total += 1
                        local_failed += 1

                async with lock:
                    total_hits += local_total
                    failure_hits += local_failed

            tasks = [asyncio.create_task(worker()) for _ in range(connections)]

            await asyncio.gather(*tasks)

        return WrkResult(
            url=api_data.url,
            method=api_data.method,
            connections=connections,
            time_taken=f"{time.time() - start:.3f}s",
            total_hits=total_hits,
            success_hits=total_hits - failure_hits,
            failure_hits=failure_hits,
            success_messages=success_codes,
            failure_messages=failure_codes,
        )